src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+    /// reqwest::MultipartField::param("key", "value").file_name(Some(file_name));
+    /// ```
+    ///
+    pub fn file_name<T: Into<Cow<'static, str>>>(mut self, filename: Option<T>) -> MultipartField {
+        self.file_name = filename.and_then(|filename| Some(filename.into()));
+        self
+    }
+    fn header(&self) -> String {
+        fn percent_encode(input: &str) -> String {
+            let mut result = String::with_capacity(input.len());
+            for character in input.chars() {
+                match character {
+                    // check for illegal ascii
+                    '%' => result.push_str("%25"),
+                    '"' => result.push_str("%22"),
+                    '\'' => result.push_str("%27"),
+                    '\r' => result.push_str("%0D"),
+                    '\n' => result.push_str("%0A"),
+                    // check for legal ascii
+                    other if other.is_ascii() => result.push(other),
+                    // must be non-ascii utf8
+                    other => {
+                        let mut buffer = [0; 4];
+                        for byte in other.encode_utf8(&mut buffer).bytes() {
+                            result.push_str(&format!("%{:X}", byte));
+                        }
+                    }
+                }
+            }
+            result
+        }
+
+        fn format_parameter(name: &str, value: &str) -> String {
+            let legal_value = percent_encode(value);
+            if value.len() == legal_value.len() {
+                // nothing has been percent encoded
+                format!("{}=\"{}\"", name, value)
+            } else {
+                // something has been percent encoded
+                format!("{}*=utf-8''{}", name, legal_value)
+            }
+        }
+
+        format!(
+            // TODO: I would use hyper's ContentDisposition header here, but it doesnt seem to have
+            // the form-data type
+            "Content-Disposition: form-data; {}{}{}",
+            format_parameter("name", self.name.as_ref()),
+            match self.file_name {
+                Some(ref file_name) => format!("; {}", format_parameter("filename", file_name)),
+                None => "".to_string(),
+            },
+            match self.mime {
+                Some(ref mime) => {
+                    format!(
+                        // TODO: Apparently I still have to write out Content-Type here?!
+                        // I thought header would format itself that way on its own
+                        "\r\nContent-Type: {}",
+                        ::header::ContentType(mime.clone())
+                    )
+                }
+                None => "".to_string(),
+            }
+        )
+    }
+}
+
+pub struct RequestReader {
+    request: MultipartRequest,
+    active_reader: Option<Box<Read + Send>>,
+}
+
+// TODO: RequestReader cannot derive debug because active_reader is not Debug
+// Not sure how to best resolve this...
```

Comment:
`RequestReader` has to be public because it is returned in the public `reader` method on `MultipartRequest`.  The only alternative I see is making `reader` return `Box<Read>`.